### ISSUE TITLE
UserRepoKeyRing lost!!!

### ISSUE THREAD

**Issue by nlmarco** [subshare/subshare#55]
Today, I encountered (fortunately personally) the following exception, even though subshare worked perfectly fine, yesterday -- and I did not change anything.

1)
```
2017-07-07 10:22:04,362 [RepoSyncDaemonThread_0_0] ERROR c.c.c.c.r.s.RepoSyncDaemonImpl$WrapperRunnable - run: java.lang.ExceptionInInitializerError
java.lang.ExceptionInInitializerError: null
        at org.subshare.core.user.UserRegistryImpl.getInstance(UserRegistryImpl.java:78) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRepoKeyRingLookupImpl.getUserRepoKeyRing(UserRepoKeyRingLookupImpl.java:34) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRepoKeyRingLookupImpl.getUserRepoKeyRing(UserRepoKeyRingLookupImpl.java:24) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.getUserRepoKeyRing(CryptreeRestRepoTransportImpl.java:644) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.determinePathPrefix(CryptreeRestRepoTransportImpl.java:200) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at co.codewizards.cloudstore.core.repo.transport.AbstractRepoTransport.getPathPrefix(AbstractRepoTransport.java:84) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.getCryptree(CryptreeRestRepoTransportImpl.java:637) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.syncCryptoKeysFromRemoteRepo(CryptreeRestRepoTransportImpl.java:244) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.getChangeSetDto(CryptreeRestRepoTransportImpl.java:168) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.sync(RepoToRepoSync.java:223) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.syncDown(RepoToRepoSync.java:174) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.sync(RepoToRepoSync.java:137) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at org.subshare.core.repo.sync.SsRepoToRepoSync.sync(SsRepoToRepoSync.java:65) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoSyncRunner.run(RepoSyncRunner.java:51) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoSyncDaemonImpl$WrapperRunnable.run(RepoSyncDaemonImpl.java:123) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_131]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_131]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_131]
Caused by: org.subshare.core.pgp.MissingSigningPgpKeyException: The data was signed using the following PGP-keys, of which none could be found in the local key-ring: [xxxxxxxxxxxxxxxx]
        at org.subshare.core.pgp.gnupg.BcPgpDecoder.verifySignature(BcPgpDecoder.java:336) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.pgp.gnupg.BcPgpDecoder.decodeCompressed(BcPgpDecoder.java:180) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.pgp.gnupg.BcPgpDecoder.decode(BcPgpDecoder.java:103) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRepoKeyImpl.verifyPublicKeyData(UserRepoKeyImpl.java:181) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRepoKeyImpl.verifyPublicKeyData(UserRepoKeyImpl.java:173) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRepoKeyImpl.<init>(UserRepoKeyImpl.java:77) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRepoKeyDtoConverter.fromUserRepoKeyDto(UserRepoKeyDtoConverter.java:33) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRepoKeyRingDtoConverter.fromUserRepoKeyRingDto(UserRepoKeyRingDtoConverter.java:27) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserDtoConverter.fromUserDto(UserDtoConverter.java:50) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRegistryImpl.readPayloadEntry(UserRegistryImpl.java:130) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.fbor.FileBasedObjectRegistry.readPayload(FileBasedObjectRegistry.java:181) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.fbor.FileBasedObjectRegistry.read(FileBasedObjectRegistry.java:68) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.fbor.FileBasedObjectRegistry.read(FileBasedObjectRegistry.java:79) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRegistryImpl.read(UserRegistryImpl.java:98) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRegistryImpl.<init>(UserRegistryImpl.java:83) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRegistryImpl$Holder.<clinit>(UserRegistryImpl.java:74) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        ... 20 common frames omitted
```

2)
```
2017-07-07 10:23:01,146 [RepoSyncDaemonThread_0_2] ERROR c.c.c.c.r.s.RepoSyncDaemonImpl$WrapperRunnable - run: java.lang.NoClassDefFoundError: Could not initialize class org.subshare.core.user.UserRegistryImpl$Holder
java.lang.NoClassDefFoundError: Could not initialize class org.subshare.core.user.UserRegistryImpl$Holder
        at org.subshare.core.user.UserRegistryImpl.getInstance(UserRegistryImpl.java:78) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRepoKeyRingLookupImpl.getUserRepoKeyRing(UserRepoKeyRingLookupImpl.java:34) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.core.user.UserRepoKeyRingLookupImpl.getUserRepoKeyRing(UserRepoKeyRingLookupImpl.java:24) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.getUserRepoKeyRing(CryptreeRestRepoTransportImpl.java:644) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.determinePathPrefix(CryptreeRestRepoTransportImpl.java:200) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at co.codewizards.cloudstore.core.repo.transport.AbstractRepoTransport.getPathPrefix(AbstractRepoTransport.java:84) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.getCryptree(CryptreeRestRepoTransportImpl.java:637) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.syncCryptoKeysFromRemoteRepo(CryptreeRestRepoTransportImpl.java:244) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.getChangeSetDto(CryptreeRestRepoTransportImpl.java:168) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.sync(RepoToRepoSync.java:223) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.syncDown(RepoToRepoSync.java:174) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.sync(RepoToRepoSync.java:137) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at org.subshare.core.repo.sync.SsRepoToRepoSync.sync(SsRepoToRepoSync.java:65) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoSyncRunner.run(RepoSyncRunner.java:51) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoSyncDaemonImpl$WrapperRunnable.run(RepoSyncDaemonImpl.java:123) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_131]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_131]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_131]
```

I then closed subshare completely (including the background process) and started it again. After starting it again, all syncs failed with this exception:

3)
```
2017-07-07 10:26:27,789 [RepoSyncDaemonThread_0_0] ERROR c.c.c.c.r.s.RepoSyncDaemonImpl$WrapperRunnable - run: java.lang.IllegalArgumentException: userRepoKeyRing.getUserRepoKeys(serverRepositoryId).isEmpty() :: serverRepositoryId=xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx
java.lang.IllegalArgumentException: userRepoKeyRing.getUserRepoKeys(serverRepositoryId).isEmpty() :: serverRepositoryId=ddddca5b-676a-43ba-9de8-1113043e37d9
        at org.subshare.local.CryptreeContext.<init>(CryptreeContext.java:76) ~[org.subshare.local-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.local.CryptreeImpl.getCryptreeContext(CryptreeImpl.java:301) ~[org.subshare.local-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.local.CryptreeImpl.putCryptoChangeSetDto(CryptreeImpl.java:536) ~[org.subshare.local-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.syncCryptoKeysFromRemoteRepo(CryptreeRestRepoTransportImpl.java:246) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at org.subshare.rest.client.transport.CryptreeRestRepoTransportImpl.getChangeSetDto(CryptreeRestRepoTransportImpl.java:168) ~[org.subshare.rest.client-0.10.1-SNAPSHOT.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.sync(RepoToRepoSync.java:223) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.syncDown(RepoToRepoSync.java:174) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoToRepoSync.sync(RepoToRepoSync.java:137) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at org.subshare.core.repo.sync.SsRepoToRepoSync.sync(SsRepoToRepoSync.java:65) ~[org.subshare.core-0.10.1-SNAPSHOT.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoSyncRunner.run(RepoSyncRunner.java:51) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at co.codewizards.cloudstore.core.repo.sync.RepoSyncDaemonImpl$WrapperRunnable.run(RepoSyncDaemonImpl.java:123) ~[co.codewizards.cloudstore.core-0.10.0.jar:na]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [na:1.8.0_131]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [na:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_131]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_131]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_131]
```

I checked the file `~/.subshare/userRegistry.subshare` and saw that my `UserDto`'s `UserRepoKeyRingDto` was completely empty!

Unfortunately, the broken `~/.subshare/userRegistry.subshare` was synced up into the locker. My private repository keys were thus lost both locally and on the server!

Fortunately, I was able to restore the `~/.subshare/userRegistry.subshare` from another computer. Down-syncing there did not cause any further data loss, because the data is merged -- thus preserving the correct key-ring.

I modified one user's name on the functional machine to make subshare sync the user registry up to the server again. Afterwards, subshare worked fine again on the computer having originally caused the problem (it synced the correct registry down automatically).

There are thus multiple issues to be solved in the context of this issue:

1) We should keep a few old versions (maybe 5 or 10) of all files in ~/.subshare/ -- most importantly the user-registry.

2) When reading a file failed, it should not be overwritten!

3) On the server, the lockers should keep multiple old versions, too.

4) The class should still be loaded. Maybe the singleton should be null or it should be marked "broken" causing method invocations to fail (same result as it being null, but maybe nicer error message). This is low priority, though.

Of course, it would be nice to find out, why the original error happened, but I doubt that this is possible due to the file having been overwritten :-(

**Comment by nlmarco** [comment]
=> exception 1.
The PGP key being named in the exception really does not exist in my OpenPGP key ring.

**Comment by nlmarco** [comment]
The same error happened lately to a friend of mine. So, now it really was time to fix it ;-)

He used multiple different OpenPGP keys -- which clearly caused this same error:

This error happened, because the `UserRegistryImpl` (more precisely the `UserRepoKeyImpl`) tried to decrypt the secret key (using the OpenPGP key) already when reading the file. If the passphrase for the OpenPGP key was not provided, this failed.

I modified this behaviour: It now lazily decrypts the secret key, when it needs it the first time.

Additionally, I introduced a backup in `FileBasedObjectRegistry` (the super-class of `UserRegistryImpl`): Whenever a file is written, a backup is created in `~/.subshare/backup/`. Old backup-files are deleted, if they are older than 2 days and if there are more than 10 backup files (per type).

I still have no idea about this unknown PGP key in the exception that happened on my machine, though. But since this actual problem is now fixed and should not re-occur (I introduced a new test: `MultiplePgpKeysSingleUserRegistryIT`), I close this issue.
